api/pkg/s3/bucketput.go

```diff
@@ -16,60 +16,75 @@ package s3
 	"github.com/opensds/multi-cloud/api/pkg/common"
 	c "github.com/opensds/multi-cloud/api/pkg/context"
-	. "github.com/opensds/multi-cloud/s3/pkg/exception"
+	"github.com/opensds/multi-cloud/s3/error"
 	"github.com/opensds/multi-cloud/s3/pkg/model"
 	"github.com/opensds/multi-cloud/s3/proto"
+	log "github.com/sirupsen/logrus"
+	"net/http"
+	"strings"
+	"time"
 )
 
 func (s *APIService) BucketPut(request *restful.Request, response *restful.Response) {
-	bucketName := request.PathParameter("bucketName")
-	log.Infof("Received request for create bucket: %s", bucketName)
+	bucketName := strings.ToLower(request.PathParameter(common.REQUEST_PATH_BUCKET_NAME))
+	if !isValidBucketName(bucketName) {
+		response.WriteError(http.StatusBadRequest, s3error.ErrInvalidBucketName)
+
+		return
+	}
+	log.Infof("received request: PUT bucket[name=%s]\n", bucketName)
+
+	if len(request.HeaderParameter(common.REQUEST_HEADER_CONTENT_LENGTH)) == 0 {
+		log.Errorf("missing content length")
+		response.WriteError(http.StatusLengthRequired, s3error.ErrMissingContentLength)
+
+		return
+	}
 
 	ctx := common.InitCtxWithAuthInfo(request)
 	actx := request.Attribute(c.KContext).(*c.Context)
 	bucket := s3.Bucket{Name: bucketName}
-	body := ReadBody(request)
 	bucket.TenantId = actx.TenantId
 	bucket.UserId = actx.UserId
 	bucket.Deleted = false
-	bucket.CreationDate = time.Now().Unix()
+	bucket.CreateTime = time.Now().Unix()
+	log.Infof("Bucket PUT: TenantId=%s, UserId=%s\n", bucket.TenantId, bucket.UserId)
 
-	if body != nil {
+	body := ReadBody(request)
+	flag := false
+	if body != nil && len(body) != 0 {
+		log.Infof("request body is not empty")
 		createBucketConf := model.CreateBucketConfiguration{}
 		err := xml.Unmarshal(body, &createBucketConf)
 		if err != nil {
-			response.WriteError(http.StatusInternalServerError, err)
+			log.Infof("unmarshal failed, body:%v, err:%v\n", body, err)
+			response.WriteError(http.StatusInternalServerError, s3error.ErrUnmarshalFailed)
 			return
-		} else {
-			backendName := createBucketConf.LocationConstraint
-			if backendName != "" {
-				log.Infof("backendName is %v\n", backendName)
-				bucket.Backend = backendName
-				client := getBackendByName(ctx, s, backendName)
-				if client == nil {
-					response.WriteError(http.StatusInternalServerError, NoSuchType.Error())
-					return
-				}
-			} else {
-				log.Info("default backend is not provided.")
-				response.WriteError(http.StatusBadRequest, NoSuchBackend.Error())
-				return
-			}
+		}
+
+		backendName := createBucketConf.LocationConstraint
+		if backendName != "" {
+			log.Infof("backendName is %v\n", backendName)
+			bucket.DefaultLocation = backendName
+			flag = s.isBackendExist(ctx, backendName)
 		}
 	}
+	if flag == false {
+		log.Errorf("default backend is not provided or it is not exist.")
+		response.WriteError(http.StatusInternalServerError, s3error.ErrGetBackendFailed)
+		return
+	}
 
 	res, err := s.s3Client.CreateBucket(ctx, &bucket)
```

Comment:
are we creating a bucket or a folder on the actual cloud backend ? say AWS

Reply:
@rhsakarpos , no, the bucket will not be created in aws after CreateBucket api is called. The bucket will be created in aws when object is put.

Reply:
that does not seem right. the bucket is first created, can be empty

Reply:
@rhsakarpos sorry for before explanation, and multi-cloud will not create bucket in aws during object put. the bucket uses by object put operation is created beforehand and is managed by backend. when the user selects a backend, he will attach a bucket to it, while bucket is created in aws beforehand.

Reply:
hi, this i think is confusing. when i add a backend, it is just a cloud provider backend, i should NOT need to give a bucket there. bucket is different from backend, should be handled by OSDS.
Thoughts ?

Reply:
@nguptaopensds and @kumarashit ? we discussed this here too...this needs to be fixed. bucket is NOT EQUAL TO folder. Backend is NOT EQUAL to bucket

Reply:
@rhsakarpos check [this](https://github.com/opensds/multi-cloud/pull/191#discussion_r338526608) please.